cpp/src/parquet/arrow/reader.cc

```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
+    END_PARQUET_CATCH_EXCEPTIONS
+    // We might return less then the requested slot (i.e. reaching an end of a file)
+    // ensure we've set all the bits here.
+    if (validity_io.values_read < number_of_slots) {
+      // + 1  because arrays lengths are values + 1
+      std::fill(length_data + validity_io.values_read + 1,
+                length_data + number_of_slots + 1, 0);
+      if (validity_io.valid_bits != nullptr &&
+          BitUtil::BytesForBits(validity_io.values_read) <
+              BitUtil::BytesForBits(number_of_slots)) {
+        std::fill(validity_io.valid_bits + BitUtil::BytesForBits(validity_io.values_read),
+                  validity_io.valid_bits + BitUtil::BytesForBits(number_of_slots), 0);
+      }
+    }
+
+    RETURN_NOT_OK(item_reader_->BuildArray(length_data[validity_io.values_read], out));
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayData> item_chunk, ChunksToSingle(**out));
+
+    std::vector<std::shared_ptr<Buffer>> buffers{
+        validity_io.null_count > 0 ? validity_buffer : std::shared_ptr<Buffer>(),
+        lengths_buffer};
+    auto data = std::make_shared<ArrayData>(
+        field_->type(),
+        /*length=*/validity_io.values_read, std::move(buffers),
+        std::vector<std::shared_ptr<ArrayData>>{item_chunk}, validity_io.null_count);
+
+    std::shared_ptr<Array> result = ::arrow::MakeArray(data);
+    RETURN_NOT_OK(result->Validate());
     *out = std::make_shared<ChunkedArray>(result);
     return Status::OK();
   }
 
   const std::shared_ptr<Field> field() override { return field_; }
 
-  const ColumnDescriptor* descr() const override { return nullptr; }
-
-  ReaderType type() const override { return LIST; }
-
  private:
   std::shared_ptr<ReaderContext> ctx_;
   std::shared_ptr<Field> field_;
-  int16_t max_definition_level_;
-  int16_t max_repetition_level_;
+  ::parquet::internal::LevelInfo level_info_;
   std::unique_ptr<ColumnReaderImpl> item_reader_;
 };
 
 class PARQUET_NO_EXPORT StructReader : public ColumnReaderImpl {
  public:
   explicit StructReader(std::shared_ptr<ReaderContext> ctx,
-                        const SchemaField& schema_field,
                         std::shared_ptr<Field> filtered_field,
-                        std::vector<std::unique_ptr<ColumnReaderImpl>>&& children)
+                        ::parquet::internal::LevelInfo level_info,
+                        std::vector<std::unique_ptr<ColumnReaderImpl>> children)
       : ctx_(std::move(ctx)),
-        schema_field_(schema_field),
         filtered_field_(std::move(filtered_field)),
-        struct_def_level_(schema_field.level_info.def_level),
-        children_(std::move(children)) {}
+        level_info_(level_info),
+        children_(std::move(children)) {
+    // There could be a mix of children some might be repeated some might not be.
+    // If possible use one that isn't since that will be guaranteed to have the least
+    // number of rep/def levels.
```

Comment:
You mean "of rep levels"? You could have arbitrarily nested structs with a lot of dep levels?

Reply:
rephrased a little bit.  There is always an equal number of repetition and definition levels for any particular leaf.